### PR TITLE
Dag Context Loop invariant code motion

### PR DESCRIPTION
**Pull request by clyben**
the test is insufficient now.

### CODE REVIEW

File: dag_in_context/src/schema.egg [hunk 1]
```diff
@@ -113,7 +113,7 @@
 (function Uop   (UnaryOp Expr) Expr)
 ; gets from a tuple. static index
 (function Get   (Expr i64) Expr)
-; (Alloc amount state_edge pointer_type)
+; (Alloc id amount state_edge pointer_type)
```
Comment (yihozhang) [inline dag_in_context/src/schema.egg:116]: Nice catch

File: dag_in_context/src/utility/util.egg [hunk 2]
```diff
@@ -55,3 +55,67 @@
       ((union (Get (Concat expr1 expr2) i)
               (Get expr2 (- i len1))))
        :ruleset always-run)
+
+
+;; This Expr size actually do the same thing as cost function.
+(function Expr-size (Expr) i64 :unextractable :merge (min old new) )
```
Comment (yihozhang) [inline dag_in_context/src/utility/util.egg:61]: Maybe leave a note here that this computes the tree size (we don't need to say this before but now we are using a DAG semantics)

File: dag_in_context/src/utility/util.egg [hunk 3]
```diff
@@ -55,3 +55,67 @@
...
+(rule ((Empty ty assum)) ((set (Expr-size (Empty ty assum)) 0))  :ruleset always-run)
+
+(rule ((= expr (Cons hd tl)) 
+       (= sum (+ (ListExpr-size tl) (Expr-size hd)))) 
+      ((set (ListExpr-size expr) sum)) :ruleset always-run)
+
+(rule ((Nil)) 
+      ((set (ListExpr-size (Nil)) 0))  :ruleset always-run)
+
+(rule ((= expr (Alloc id e state ty)) ;; do state edge's expr should be counted?
```
Comment (yihozhang) [inline dag_in_context/src/utility/util.egg:119]: I'm open to this

File: dag_in_context/src/schedule.egg [hunk 4]
```diff
@@ -0,0 +1,37 @@
+(run-schedule
```
Comment (yihozhang) [inline dag_in_context/src/schedule.egg:1]: Why do we need this file and why is it a new file?

**Review by yihozhang — COMMENTED** [review]
The optimization is cleanly implemented. A few TODOs before we merge:

* Run this over the benchmark to see if the algorithm scales. Ideally, we should find cases where we can actually improve the performance on the benchmark.
* There are some new conflicts that need to be resolved.

File: dag_in_context/src/greedy_dag_extractor.rs [hunk 5]
```diff
@@ -536,8 +538,10 @@ impl CostModel for DefaultCostModel {
             "HasType" | "HasArgType" | "ContextOf" | "NoContext" | "ExpectType" => 0.,
             "ExprIsPure" | "ListExprIsPure" | "BinaryOpIsPure" | "UnaryOpIsPure" => 0.,
             "BodyContainsExpr" | "ScopeContext" => 0.,
-            "Region" | "Full" | "IntB" | "BoolB" => 0.,
+            "Region" | "Full" | "IntB" | "BoolB" | "boundary-Expr" => 0.,
```
Comment (oflatt) [inline dag_in_context/src/greedy_dag_extractor.rs:541]: I think Region and Full are gone

File: dag_in_context/src/optimizations/loop_invariant.egg [hunk 6]
```diff
@@ -25,3 +25,64 @@
...
+
+;; mock function
+(ruleset loop-inv-motion)
+
+(rule ((boundary-Expr loop inv)
+       (> (Expr-size inv) 1)
+       ;; TODO: replace Expr-size when cost model is ready
+       (= loop (DoWhile in pred_out))
+       ;; the outter assumption of the loop 
+       ;; Question: what if there are multiple loop_ctx?
```
Comment (oflatt) [inline dag_in_context/src/optimizations/loop_invariant.egg:62]: There can't be multiple contexts for one node, it's an invariant

File: dag_in_context/src/schedule.egg (hunk 4 above)
Comment (clyben) [reply dag_in_context/src/schedule.egg:1]: forget to delete this

File: dag_in_context/src/utility/util.egg [hunk 7]
```diff
@@ -55,3 +55,67 @@
       ((union (Get (Concat expr1 expr2) i)
               (Get expr2 (- i len1))))
        :ruleset always-run)
+
+
+;; This Expr size actually do the same thing as cost function.
```
Comment (oflatt) [inline dag_in_context/src/utility/util.egg:60]: please move to new file for expr size

**Review by oflatt — CHANGES_REQUESTED** [review]
Great work! Some small changes

**Review by oflatt — APPROVED** [review]
Great work!

**Pull request merged**